packages/studio/src/store/slices/pages/createPageSlice.ts

```diff
@@ -160,6 +160,79 @@ export const createPageSlice: SliceCreator<PageSlice> = (set, get) => {
         (component) => component.uuid === activeComponentUUID
       );
     },
+    addSelectedComponentUUID: (selectedUUID: string) => {
+      set((store) => {
+        store.selectedComponentUUIDs.add(selectedUUID);
+      });
+    },
+    addSelectedComponentRect: (
+      selectedUUID: string,
+      rect: DOMRectProperties
+    ) => {
+      set((store) => {
+        store.selectedComponentRectsMap[selectedUUID] = rect;
+      });
+    },
+    clearSelectedComponents: () => {
+      set((store) => {
+        store.selectedComponentUUIDs = new Set();
+        store.selectedComponentRectsMap = {};
+      });
+    },
+    /**
+     * Adds all of the components and their children ordered between the selected component and
+     * the active component (inclusive) to selectedComponentUUIDs.
+     */
+    addShiftSelectedComponentUUIDs: (selectedComponent: ComponentState) => {
+      const selectedUUID = selectedComponent.uuid;
+      const {
+        activeComponentUUID,
+        getActivePageState,
+        addSelectedComponentUUID,
+      } = get();
+      const activePageState = getActivePageState();
+      const componentTree = activePageState?.componentTree;
+      if (selectedUUID === activeComponentUUID) return;
+      if (!componentTree) {
+        throw new Error(
+          `Error adding components: component tree is undefined.`
+        );
+      }
+      if (!activeComponentUUID) {
+        throw new Error(
+          `Error adding components: active component is undefined.`
+        );
+      }
+      const targetComponentUUIDs = [selectedUUID, activeComponentUUID] as const;
+      const lowestParentUUID = ComponentTreeHelpers.getLowestParentUUID(
+        ...targetComponentUUIDs,
+        componentTree
+      );
+
+      let selecting = false;
+      ComponentTreeHelpers.mapComponentTreeParentsFirst(componentTree, (c) => {
+        if (c.parentUUID !== lowestParentUUID) return;
+        const descendants = ComponentTreeHelpers.getDescendants(
```

Comment:
I think the logic in this handler (lines 215-233) can be simplified. Based on my understanding, there are really only a few cases here:
- _the sub-tree starting with the current component includes **both** the targets_: select the whole sub-tree and exit
- _the sub-tree starting with the current component includes **one** target_: select the whole sub-tree and begin/end selecting
- _the sub-tree starting with the current component includes **no** targets_: only select the whole sub-tree if we're in the middle of selecting

Reply:
Done